dosbox-staging/dosbox-staging
# Keyboard layout autodetection fixes/improvements

# Description

1. Default keyboard layout is now `us`, not `auto`
2. Keyboard layout is now only detected if user sets `keyboard_layout = auto`, otherwise the code is not executed
3. Fix for keyboard layout detection on KDE if some of the configured layouts has variants (like `us intl`) and some do not
4. Improve the keyboard layout selection algorithm
a) host->guest layout mappings which are not good matches are now marked as _fuzzy_ and they get the lowest priority during selection
b) if the OS can provide a keyboard layout list sorted by priority (according to the user), these priorities are generally respected (this applies to Linux only)

Remarks:
- I have decided to keep the sorting behavior of giving plain English layouts a very low priority; since the US English layout is now the default, if someone decides to change it to `auto`, he is probably interested in trying something else. Moreover, on modern Windows it's the OS who decides which keyboard layouts are listed in the GUI (probably based on the installed language packs, or something similar), so US English is probably always on the list even if the user prefers the national one.
- I really don't want to spend weeks tweaking the selection algorithm; there is probably no way to make it work for everyone. Please consider the autodetection as an option for all the folks who are not satisfied with US English, but do not want to read through the list of layouts, code pages, FreeDOS documentation, etc. For some of them it will work great, for others... not really. Such is life.


## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/4076


# Release notes

Keyboard layout now defaults to `us` as a safe option for almost everyone. If you want to experiment, the `auto` option is still there, but you have set it explicitly.

# Manual testing

- Set `keyboard_layout` to some improper value; it should load US keyboard (you can check this by executing `KEYB` command).
- On Linux with KDE set two keyboard layouts - _Polish_ and _US International with dead keys_. Set `keyboard_layout = auto`. DOSBox now sets whichever keyboard layout is first on the list in the _KDE System Settings_ and the log shows the _intl_ modifier near the _us_ host value:

```
LOCALE: Keyboard layout and code page detected from 'KDE: pl us:intl'
```

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
